core/shared/src/main/scala/zio/prelude/PartialOrd.scala

```diff
@@ -0,0 +1,955 @@
+   * Since this takes a separate `equal0`, short-circuiting the equality check (failing fast) is possible.
+   */
+  def makeFrom[A](ord: (A, A) => Option[Ordering], equal0: Equal[A]): PartialOrd[A] =
+    new PartialOrd[A] {
+      override protected def checkCompareOption(l: A, r: A): Option[Ordering] = ord(l, r)
+      override protected def checkEqual(l: A, r: A): Boolean                  = equal0.equal(l, r)
+    }
+
+  /**
+   * Derives an `PartialOrd[Chunk[A]]` given an `PartialOrd[A]`.
+   */
+  implicit def ChunkPartialOrd[A: PartialOrd]: PartialOrd[Chunk[A]] =
+    makeFrom(
+      { (l, r) =>
+        val j           = l.length
+        val k           = r.length
+        val PartialOrdA = PartialOrd[A]
+
+        @tailrec
+        def loop(i: Int): Option[Ordering] =
+          if (i == j && i == k) Some(Ordering.Equals)
+          else if (i == j) Some(Ordering.LessThan)
+          else if (i == k) Some(Ordering.GreaterThan)
+          else
+            PartialOrdA.compareOption(l(i), r(i)) match {
+              case Some(Ordering.Equals) => loop(i + 1)
+              case compareOption         => compareOption
+            }
+
+        loop(0)
+      },
+      Equal.ChunkEqual
+    )
+
+  /**
+   * Derives an `PartialOrd[F[A]]` given a `Derive[F, PartialOrd]` and an `PartialOrd[A]`.
+   */
+  implicit def DerivePartialOrd[F[_], A](implicit derive: Derive[F, PartialOrd], ord: PartialOrd[A]): PartialOrd[F[A]] =
+    derive.derive(ord)
+
+  /**
+   * Derives an `PartialOrd[Either[A, B]]` given an `PartialOrd[A]` and an `PartialOrd[B]`.
+   */
+  implicit def EitherPartialOrd[A: PartialOrd, B: PartialOrd]: PartialOrd[Either[A, B]] =
+    makeFrom(
+      {
+        case (Left(a1), Left(a2))   => a1 =??= a2
+        case (Left(_), Right(_))    => Some(Ordering.LessThan)
+        case (Right(_), Left(_))    => Some(Ordering.GreaterThan)
+        case (Right(b1), Right(b2)) => b1 =??= b2
+      },
+      Equal.EitherEqual
+    )
+
+  /**
+   * Derives an `PartialOrd[List[A]]` given an `PartialOrd[A]`.
+   */
+  implicit def ListPartialOrd[A: PartialOrd]: PartialOrd[List[A]] = {
+    val PartialOrdA = PartialOrd[A]
+
+    @tailrec
+    def loop(left: List[A], right: List[A]): Option[Ordering] =
+      (left, right) match {
+        case (Nil, Nil)           => Some(Ordering.Equals)
+        case (Nil, _)             => Some(Ordering.LessThan)
+        case (_, Nil)             => Some(Ordering.GreaterThan)
+        case (h1 :: t1, h2 :: t2) =>
+          PartialOrdA.compareOption(h1, h2) match {
+            case Some(Ordering.Equals) => loop(t1, t2)
+            case compareOption         => compareOption
+          }
+      }
+
+    makeFrom((l, r) => loop(l, r), Equal.ListEqual)
+  }
+
+  /**
+   * Derives an `PartialOrd[Map[A, B]]` given an `PartialOrd[B]`.
+   * Due to the limitations of Scala's `Map`, this uses object equality on the keys.
+   */
+  implicit def MapPartialOrd[A, B: PartialOrd]: PartialOrd[Map[A, B]] =
```

Comment:
I'm not sure this definition makes sense in general.

It seems to be saying that one map is smaller than another if the keys are a subset and the values are less than or equal. This makes a lot of sense when the values represent some notion of "how many" of each key exist in the map as in `ZSet`. For example:

```scala
val shoppingList1 = Map("Apple" -> 2)
val shoppingList2 = Map("Apple" -> 3, "Orange" -> 1)
shoppingList1 < shoppingList2
```

But consider something like:

```scala
val nickNames1 = Map("William" -> "Will")
val nickNames2 = Map("William" -> "Bill")
```

Here it doesn't really seem like it makes sense to say one of these mappings is "smaller" than the other, they are just different.

I would be inclined to use a definition of subset for maps in general more analogous to sets (tuples of keys and values in one map are a subset of tuples in the other) and use a definition like the current one in `ZSet` where we have a more specific structure regarding what the values represent.

Reply:
> Here it doesn't really seem like it makes sense to say one of these mappings is "smaller" than the other, they are just different.

But that's precisely the use case for `PartialOrd`, isn't it? Neither `Map("William" -> "Will")` nor `Map("William" -> "Bill")` is less nor greater than each other, let alone being equal. That's why the comparison for them will return `None` (aka `NoOrder`, or `Incomparable` as suggested above).

I think that's an intuitive and practically useful way for `compareOption` to work.

Reply:
Right. There is definitely a certain logic embodied in the current approach, I'm just not sure it makes the most sense in general. As that example shows the ordering on the values intuitively doesn't seem to be related to the ordering of the maps. Of course if you take the current logic you can work backwards from it but your initial intuition was that these are things that are just different but not "bigger" or "smaller".

An alternative would be to implement a more limited notion of partial ordering where one map is only "smaller" than another if it contains a subset of the bindings in the other map. So `Map("William" -> "Will")` would be smaller than `Map("William" -> "Will", "Richard" -> "Dick")` but `Map("William" -> "Bill")` would not be comparable to `Map("William" -> "Bill")`. We could still use the current definition for `ZSet` where I think your implementation makes a ton of sense.

Reply:
I've change the implementation in Map's PartialOrd :+1: please have another look